### ISSUE TITLE
Slower Response on large context size

### ISSUE THREAD

**Issue by iamirulofficial** [ggerganov/llama.cpp#1515]
Hi folks, this is not really a issue, I need sort of suggestion or may be discussions , I am giving a large input , I am offloading layers to GPU here is my system output:

```
llama_model_load_internal: format     = ggjt v2 (latest)
llama_model_load_internal: n_vocab    = 32001
llama_model_load_internal: n_ctx      = 2048
llama_model_load_internal: n_embd     = 5120
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 40
llama_model_load_internal: n_layer    = 40
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 13824
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 13B
llama_model_load_internal: ggml ctx size =  90.75 KB
llama_model_load_internal: mem required  = 9807.49 MB (+ 1608.00 MB per state)
llama_model_load_internal: [cublas] offloading 40 layers to GPU
llama_model_load_internal: [cublas] offloading output layer to GPU
llama_model_load_internal: [cublas] total VRAM used: 7660 MB
llama_init_from_file: kv self size  = 1600.00 MB

system_info: n_threads = 2 / 2 | AVX = 1 | AVX2 = 1 | AVX512 = 1 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 | 
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.800000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 2048, n_batch = 512, n_predict = 324, n_keep = 0
```
And time taken output
```
llama_print_timings:        load time = 22627.53 ms
llama_print_timings:      sample time =   154.86 ms /   259 runs   (    0.60 ms per token)
llama_print_timings: prompt eval time = 21530.18 ms /  1024 tokens (   21.03 ms per token)
llama_print_timings:        eval time = 125551.93 ms /   258 runs   (  486.64 ms per token)
llama_print_timings:       total time = 159984.71 ms
```

Takes around 2min30sec to complete the output , is there any way to make it under lets say 1 min or 30 sec ?

Thanks :)

**Comment by iamirulofficial** [comment]
I am running like this
```
./main --gpu-layers 1000 -m /models/gpt4-x-vicuna-13B.ggml.q4_0.bin -n 324 -c 2048  -p "This is a large prompt"
```

**Comment by SlyEcho** [comment]
```
n_threads = 2 / 2
```

Try using more cores. Is this a virtual machine? Try at least 8 cores.

**Comment by iamirulofficial** [comment]
Does threads really matters if I am offloading all layers to GPU ?

**Comment by KerfuffleV2** [comment]
Are you saying it is _immediately_ slower when you specify a larger context size, or it gradually slows down as tokens are generated and the available context fills up?

If the latter, then I'm pretty sure that's just a limitation of current technology. In fact, the amount of computation doesn't even increase linearly. This is in large part why most available models have a context limit of 2048 or lower: training (and to a somewhat lesser degree inference) just becomes prohibitively slow.

**Comment by SlyEcho** [comment]
I get 200 ms to generate tokens with my CPU with 4 threads. So it is either not using the GPU for you or there is some other issue.

How did you build the program? Did you try `make clean` first when using Make? If you use CMake did you accidentally enable Debug mode?

**Comment by iamirulofficial** [comment]
I just followed build with cublas method mentioned 
```
llama_model_load_internal: [cublas] offloading 40 layers to GPU
llama_model_load_internal: [cublas] offloading output layer to GPU
llama_model_load_internal: [cublas] total VRAM used: 7660 MB
```
I am able to get this, I can see the resources stats that it is using GPU, and without GPU it is very slow in the same machine, not sure where is the issue, I will try with CPU only once

**Comment by SlyEcho** [comment]
> I can see the resources stats that it is using GPU

I can see that as well because `21.03 ms per token` can not be done with 2 CPU threads, only with cuBLAS.

You didn't mention the GPU itself, but I assume it has at least 8 GB of VRAM, so it is probably not an integrated GPU. (I found that integrated GPU's are not usable for this feature, at least AMD ones)

**Comment by KerfuffleV2** [comment]
@iamirulofficial It's still not clear what you're saying is slower.

You might mean just changing `-c` instantly makes a difference.

You might mean that as tokens are generated, each subsequent token takes more time.

You might even just be talking about how feeding the prompt is faster than generating tokens after that point.

It's hard to help you without knowing specifically what you're saying is slow and under what conditions.

**Comment by KerfuffleV2** [comment]
@iamirulofficial 

> The overall time taken by the process including eval + completion is around 2 min 36 sec, I wanted to discuss about how to make that under 30 to 45 sec (Eval + Gen)

You still didn't answer the questions I asked.

If you're just asking how to _generally_ make it go faster and not implying you think there's any problem/unexpected behavior with GGML or llama.cpp then you basically have three options:

1. Upgrade your hardware.
2. Pay Mr. GG or other developers to work on optimizing stuff so it runs faster.
3. Wait and see if the normal rate of progress/improvement (which is actually really fast in this project) hits your target eventually.

For the second two, you're probably not going to see it go from 2 1/2 minutes to 30sec though. I mean, you're already running it fully on GPU and people have already done a lot to optimize performance. Expecting 4x increased performance from only software changes is probably pretty unrealistic.

**Comment by iamirulofficial** [comment]
Okay, I just *only* wanted a discussion whether with some or other thing it can be made faster, upgrading hardware or whether it can be made faster and I am doing something wrong, as simple as that, in the first comment itself I have said this 
``` this is not really a issue, I need sort of suggestion or may be discussions ``` if you see,

**Comment by KerfuffleV2** [comment]
> this is not really a issue, I need sort of suggestion or may be discussions

Apologies, I did miss that part but it doesn't really change a lot.

> or whether it can be made faster and I am doing something wrong

How can I can answer that question in a specific way when you didn't describe the problem/conditions in a specific way?

Just for example:

`-c 64` - Now the context size is small.

`-c 2048` - Now the context size is large.

Then we use a short prompt `-p 'I like pie'` and have the model generate 10 tokens with `-n 10`. Taking what you said literally, in the second situation ("large context size") it's slow now.

Is that what you actually meant when you said "context size"? I don't know: and you didn't answer the question I asked that was trying to get you to be more specific.

You might be saying "context size" when you don't actually mean "context ***size*** (limit)" but are talking just using a large prompt and then generating a bunch of tokens.

Before your last response or so, it also wasn't clear if you were talking about a relative change like "token 1,000 is generated slower than token 10".

By the way, a 4th option occurred to me: Go down from a 13b model to a ~3-4B one. That's another way to get an immediate ~4x speed increase. It's possible you may be able to find a model tuned specifically for your use case (text summaries) which has equivalent quality for that use case even at a smaller parameter count. I don't personally have any recommendations, so you'd have to investigate that possibility yourself.

**Comment by SlyEcho** [comment]
> The overall time taken by the process including eval + completion is around 2 min 36 sec, I wanted to discuss about how to make that under 30 to 45 sec (Eval + Gen)

The evaluation takes 21.5 seconds, this is not a problem, it seems cuBLAS can handle this fine.

The generation takes 125.5 seconds, this is the problem. I can get 2x performance than that using CPU alone (4 threads), with the GPU it is 2x faster still.

> I am using a 15 GB T4 GPU machine, with only 2 threads

It's possible that the kernels are not optimized for this hardware, as we are only testing with consumer GPUs (thoughts, @JohannesGaessler?)
In this case it would be better for you to use 4, 8, or more CPU threads to generate the output, you don't need much VRAM in this case, either.

**Comment by JohannesGaessler** [comment]
The hardware used seems to provide a lot of tensor cores. However, the implementation that I did does not utilize tensor cores and since it's matrix vector multiplication I think it's not possible either since tensor cores would only be useful once you were to process 8 vectors at once. Also keep in mind that large parts of the ggml graph currently are still GPU only; with "full" GPU offloading with my RTX 3090 the CPU still makes up ~1/3 of the runtime. I don't know with certainty what's happening either and honestly I don't want to spend my time debugging the setups of other people when there are still big TODOs that I know need to be fixed for optimal performance.
